llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+	{
+		// Tell this Module to materialize everything and release the GVMaterializer.
+		if (Error Err = M->materializeAll()) {
+			std::string Msg;
+			handleAllErrors(std::move(Err), [&](ErrorInfoBase &EIB) {
+					Msg = EIB.message();
+					});
+			if (ErrStr)
+				*ErrStr = Msg;
+			// We got an error, just return 0
+			return nullptr;
+		}
+
+		return new PartialInterpreter(std::move(M));
+	}
+};
+
+static void removeEdgeBetweenBlocks(llvm::BasicBlock* from, llvm::BasicBlock* to)
+{
+	//Given two BasicBlocks, remove all edges between them.
+	//Multiple edges are legal and might happen since SwitchInst might have more that one case going to the same BB
+
+	{
+		//First count how many edges there are betwen from and to
+		uint32_t cardinality = 0;
+		for (BasicBlock* bb : predecessors(to))
+			if (bb == from)
+				cardinality++;
+
+		//Then call removePredecessor to fix up the phis of to
+		for (uint32_t i=0; i<cardinality; i++)
+			to->removePredecessor(from);
+	}
+
+	{
+		//Then fix up from's terminator
+		llvm::Instruction* I = from->getTerminator();
+
+		switch (I->getOpcode())
+		{
+			case Instruction::Br:
+			{
+				BranchInst* BI = cast<BranchInst>(I);
+				if (BI->isConditional() && BI->getSuccessor(0) != BI->getSuccessor(1))
+				{
+					//Turn to unconditional
+					llvm::BasicBlock* other = nullptr;
+					if (BI->getSuccessor(0) != to)
+						other = BI->getSuccessor(0);
+					else
+						other = BI->getSuccessor(1);
+					BI->eraseFromParent();
+					BranchInst::Create(other, from);
+				}
+				else
+				{
+					BI->eraseFromParent();
+					new UnreachableInst(from->getParent()->getParent()->getContext(), from);
+				}
+				return;
+			}
+			case Instruction::Switch:
+			{
+				SwitchInst* SI = cast<SwitchInst>(I);
+				for (SwitchInst::CaseIt i = SI->case_end(), e = SI->case_begin(); i != e;)
+				{
+					--i;
+					auto *Successor = i->getCaseSuccessor();
+					if (Successor == to) {
+						SI->removeCase(i);
+					}
+				}
+
+				if (SI->getDefaultDest() == to)
+				{
+					if (SI->getNumSuccessors() == 1)
+					{
+						SI->eraseFromParent();
+						new UnreachableInst(from->getParent()->getParent()->getContext(), from);
+						return;
+					}
+					
+					BasicBlock* another = nullptr;
+					for (SwitchInst::CaseIt i = SI->case_end(), e = SI->case_begin(); i != e;)
+					{
+						--i;
+						auto *Successor = i->getCaseSuccessor();
+						another = Successor;
+						SI->removeCase(i);
+						break;
+					}
+					assert(another);
+					SI->setDefaultDest(another);
+				}
+				return;
+			}
+			case Instruction::Invoke:
+			{
+				//Given that we currently do not treat invokes the only way to get here would be via having the Invoke itself being unreachable
+				BasicBlock* BB = I->getParent();
+				I->replaceAllUsesWith(UndefValue::get(I->getType()));
+				I->eraseFromParent();
+				new UnreachableInst(BB->getParent()->getParent()->getContext(), BB);
+				return;
+			}
+			case Instruction::Unreachable:
+			{
+				return;
```

Comment:
This can never happen?

Reply:
It might have been generated by a previous iteration, in degenerate cases.